solutions/azure-hub-spoke-connected-group/bicep/modules/avnm.bicep

```diff
@@ -0,0 +1,315 @@
+  properties: {
+    roleDefinitionId: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b24988ac-6180-42a0-ab88-20f7382dd24c') // Contributor: b24988ac-6180-42a0-ab88-20f7382dd24c
+    principalId: userAssignedIdentity.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}
+
+@description('This is the securityadmin configuration assigned to the AVNM')
+resource securityConfig 'Microsoft.Network/networkManagers/securityAdminConfigurations@2022-05-01' = {
+  name: 'sg-${location}'
+  parent: networkManager
+   properties: {
+    applyOnNetworkIntentPolicyBasedServices: ['None'] 
+    description: 'Security Group for AVNM'
+     }
+  }
+
+@description('This is the rules collection for the security admin config assigned to the AVNM')
+resource rulesCollection 'Microsoft.Network/networkManagers/securityAdminConfigurations/ruleCollections@2022-05-01' = {
+name: 'rc-${location}'
+parent: securityConfig
+properties: {
+  appliesToGroups: [
+     {
+      networkGroupId: networkManager::networkGroupAll.id
+     }
+   ]
+ }
+}
+
+@description('This is the rule is to Deny all TCP Rules')
+resource rule1 'Microsoft.Network/networkManagers/securityAdminConfigurations/ruleCollections/rules@2022-05-01' = if (deployDefaultDenySecurityAdminRules) {
+name: 'r-tcp-${location}'
+kind: 'Custom'
+parent: rulesCollection
+properties: {
+  access: 'Deny'
+  description: 'Denying TCP rules'
+  destinationPortRanges: ['20','21','22','23','69','119','161','445','512','514','873','3389','5800','5900']
+  destinations: [
+          {
+            addressPrefix: '*'
+            addressPrefixType: 'IPPrefix'
+          }
+      ]
+  direction: 'Inbound'
+  priority: 100
+  protocol: 'TCP'
+  sourcePortRanges: ['0-65535']
+  sources: [
+          {
+            addressPrefix: '*'
+            addressPrefixType: 'IPPrefix'
+          }
+      ]  
+  }      
+}
+
+@description('This rule to deny inbound TCP/UDP rules')
+resource rule2 'Microsoft.Network/networkManagers/securityAdminConfigurations/ruleCollections/rules@2022-05-01' = {
+name: 'r-tcp-udp-${location}'
+kind: 'Custom'
+parent: rulesCollection
+properties: {
+  access: 'Deny'
+  description: 'Deny all TCP/UDP rules'
+  destinationPortRanges: ['11','135','162','593','2049']
+  destinations: [
+          {
+            addressPrefix: '*'
+            addressPrefixType: 'IPPrefix'
+          }
+      ]
+  direction: 'Inbound'
+  priority: 101
+  protocol: 'TCP,UDP'
+  sourcePortRanges: ['0-65535']
+  sources: [
+          {
+            addressPrefix: '*'
+            addressPrefixType: 'IPPrefix'
+          }
+      ]  
+  }      
+}
+
+@description('This rule to deny inbound UDP rules')
+resource rule3 'Microsoft.Network/networkManagers/securityAdminConfigurations/ruleCollections/rules@2022-05-01' = {
+name: 'r-udp-${location}'
+kind: 'Custom'
+parent: rulesCollection
+properties: {
+  access: 'Deny'
+  description: 'Deny all UDP rules'
+  destinationPortRanges: ['69','11211']
```

Comment:
Another "deny all" but not really.

Any chance we can instead just do a real DENY ALL TCP/UDP and then show an example of an "Allow X"

Reply:
I added a couple new rules to demo 'allow' (Azure SQL) and 'alwaysallow' (Azure AD). A deny all by AVNM seems like a heavy hammer to me, since then you are needing to ensure that you're allowing all required traffic for any service. Happy to discuss...

```bicep
@description('This example rule always allows outbound traffic to Azure Active Directory, overriding NSG outbound restrictions')
resource rule4 'Microsoft.Network/networkManagers/securityAdminConfigurations/ruleCollections/rules@2022-05-01' = {
  name: 'r-alwaysallow-${location}'
  kind: 'Custom'
  parent: rulesCollection
  properties: {
    access: 'AlwaysAllow'
    description: 'Always allow outbound traffic to Azure Active Directory'
    destinationPortRanges: [ '0-65535' ]
    destinations: [
      {
        addressPrefix: 'AzureActiveDirectory'
        addressPrefixType: 'ServiceTag'
      }
    ]
    direction: 'Outbound'
    priority: 103
    protocol: 'Any'
    sourcePortRanges: [ '0-65535' ]
    sources: [
      {
        addressPrefix: '*'
        addressPrefixType: 'IPPrefix'
      }
    ]
  }
}

@description('This example rule allows outbound traffic to Azure SQL, unless an NSG in the path denies it')
resource rule5 'Microsoft.Network/networkManagers/securityAdminConfigurations/ruleCollections/rules@2022-05-01' = {
  name: 'r-allow-${location}'
  kind: 'Custom'
  parent: rulesCollection
  properties: {
    access: 'Allow'
    description: 'Allow outbound traffic to Azure SQL'
    destinationPortRanges: [ '0-65535' ]
    destinations: [
      {
        addressPrefix: 'Sql'
        addressPrefixType: 'ServiceTag'
      }
    ]
    direction: 'Outbound'
    priority: 104
    protocol: 'Any'
    sourcePortRanges: [ '0-65535' ]
    sources: [
      {
        addressPrefix: '*'
        addressPrefixType: 'IPPrefix'
      }
    ]
  }
```

Reply:
deny all + allow is how Azure Firewall Works.  And it's how NSG rules best practices list (only in that case it's, "allow x,y,z then add a deny all").  We mirror that in our guidance where ever possible.  This is a prime example of where it would be possible.